dspace-api/src/main/java/org/dspace/google/GoogleAsyncEventListener.java

```diff
@@ -0,0 +1,236 @@
+                    log.error("Event information: " + ue);
+                    Context context = ue.getContext();
+                    if (context != null) {
+                        log.error("Context information:");
+                        log.error("    Current User: " + context.getCurrentUser());
+                        log.error("    Extra log info: " + context.getExtraLogInfo());
+                        if (context.getEvents() != null && !context.getEvents().isEmpty()) {
+                            for (int x = 1; x <= context.getEvents().size(); x++) {
+                                log.error("    Context Event " + x + ": " + context.getEvents().get(x));
+                            }
+                        }
+                    } else {
+                        log.error("UsageEvent has no Context object");
+                    }
+                }
+            }
+        }
+    }
+
+    private String getObjectName(UsageEvent ue) {
+        try {
+            if (ue.getObject().getType() == Constants.BITSTREAM) {
+                // For a bitstream download we really want to know the title of the owning item
+                // rather than the bitstream name.
+                return ContentServiceFactory.getInstance().getDSpaceObjectService(ue.getObject())
+                        .getParentObject(ue.getContext(), ue.getObject()).getName();
+            } else {
+                return ue.getObject().getName();
+            }
+        } catch (SQLException e) {
+            // This shouldn't merit interrupting the user's transaction so log the error and continue.
+            log.error("Error in Google Analytics recording - can't determine ParentObjectName for bitstream " +
+                    ue.getObject().getID(), e);
+        }
+
+        return null;
+
+    }
+
+    private String getIPAddress(HttpServletRequest request) {
+        String clientIP = request.getRemoteAddr();
+        if (configurationService.getBooleanProperty("useProxies", false) &&
+                request.getHeader("X-Forwarded-For") != null) {
+            /* This header is a comma delimited list */
+            for (String xfip : request.getHeader("X-Forwarded-For").split(",")) {
+                /* proxy itself will sometime populate this header with the same value in
+                    remote address. ordering in spec is vague, we'll just take the last
+                    not equal to the proxy
+                */
+                if (!request.getHeader("X-Forwarded-For").contains(clientIP)) {
+                    clientIP = xfip.trim();
+                }
+            }
+        }
+
+        return clientIP;
+    }
+
+    @PreDestroy
+    public void destroy() throws InterruptedException {
+        destroyed = true;
+        if (StringUtils.isNotEmpty(analyticsKey)) {
+            future.cancel(true);
+            executor.shutdown();
+            executor.awaitTermination(1, TimeUnit.SECONDS);
```

Comment:
Do I understand correctly that, in the event of DSpace being shutdown. This service will attempt to tidy up and post the remaining hits to GA for 1 second. If unsuccessful then those statistics are lost?

Just for curiosity's sake, there is a similar module called IRUS for DSpace with a slightly different design whereby unsuccessful hits are logged in a database table and a cron-job attempts to post them. Pros and cons there.

See https://github.com/atmire/IRUS/tree/stable_6x